File: mobile_client/lib/services/friend_service.dart
import 'dart:async';

import 'package:flutter/foundation.dart';
import 'package:mobile_client/constants/socket_events/friend_event.dart';
import 'package:mobile_client/constants/socket_events/notification_event.dart';
import 'package:mobile_client/models/friend_lists.dart';
import 'package:mobile_client/models/friend_update.dart';
import 'package:mobile_client/models/notification.dart';
import 'package:mobile_client/services/notification_service.dart';
import 'package:mobile_client/services/websocket_service.dart';

class FriendService extends ChangeNotifier {
  final WebSocketService _webSocketService;
  final NotificationService _notificationService;

  List<String> friendList = [];
  List<String> notFriendList = [];
  List<String> pendingList = [];

  FriendService(this._webSocketService, this._notificationService);

  void initialize() {
    _webSocketService.on<dynamic>(FriendEvent.updateSingleFriend.name, (data) {
      final update = FriendUpdate.fromJson(data);
      updateFriend(update.username, update.isAdded);
    });
    _webSocketService.on<String>(FriendEvent.requestIgnored.name, (username) {
      requestIgnored(username);
    });
    _webSocketService.on<dynamic>(NotificationEvent.addNotification.name,
        (data) {
      final notification = Notification.fromJson(data);
      notFriendList.remove(notification.senderUsername);
    });
    _webSocketService.on<String>(FriendEvent.addNotFriend.name, (username) {
      notFriendList.add(username);
      notifyListeners();
    });
    getFriendsEvent();
  }

  void removeListeners() {
    _webSocketService.removeAllListeners(FriendEvent.updateSingleFriend.name);
    _webSocketService.removeAllListeners(FriendEvent.requestIgnored.name);
  }

  Future<void> getFriendsEvent() async {
    final data = await _webSocketService
        .sendWithAck<void, dynamic>(FriendEvent.getFriends.name);

    friendList.clear();
    notFriendList.clear();
    pendingList.clear();

    final lists = FriendLists.fromJson(data);
    friendList = lists.friends;
    notFriendList = lists.notFriends;
    pendingList = lists.pending;

    notifyListeners();
  }

  void removeFriend(String username) {
    _webSocketService.send(FriendEvent.removeFriend.name, username);
  }

  void sendFriendRequest(String username) {
    _webSocketService.send(NotificationEvent.sendFriendRequest.name, username);
    pendingList.add(username);
    notFriendList.remove(username);
    notifyListeners();
  }

  void updateFriend(String username, bool isAdded) {
    if (isAdded) {
      friendList.add(username);
      pendingList.remove(username);
    } else {
      notFriendList.add(username);
      friendList.remove(username);
    }
    notifyListeners();
  }

  void requestIgnored(String username) {
    pendingList.remove(username);
    notFriendList.add(username);
    notifyListeners();
  }

  String? validateEntry(value) {
    if (value == null || value.isEmpty) {
      return "Veuillez entrer le nom d'un utilisateur";
    }

    value = value.toString().trim();
    if (notFriendList.contains(value)) {
      return null;
    }
    if (friendList.contains(value)) {
      return "Vous êtes déjà des amis";
    }
    if (pendingList.contains(value)) {
      final notificationExists = _notificationService.notificationList.any(
        (notif) => notif.senderUsername == value,
      );
      if (notificationExists) {
        return "Vous avez une demande de cet utilisateur";
      }
      return "Vous avez déjà envoyé une demande à cet utilisateur";
    }
    return "Entrez un nom d'utilisateur valide";
  }
}

File: desktop-client/src/app/account/services/friend/friend.service.ts
import { Injectable, OnDestroy } from '@angular/core';
import { FriendRequestResponse } from '@app/account/interfaces/friend-request-response';
import { WebsocketService } from '@app/shared/services/websocket/websocket.service';
import { NOT_FOUND_INDEX } from '@common/constants/verif-constants';
import { FriendLists } from '@common/interfaces/friend-lists';
import { FriendUpdate } from '@common/interfaces/friend-update';
import { Notification } from '@common/interfaces/notification';
import { FriendEvent } from '@common/socket-events/friend-event';
import { NotificationEvent } from '@common/socket-events/notification-event';

@Injectable({
    providedIn: 'root',
})
export class FriendService implements OnDestroy {
    friendList: string[] = [];
    notFriendList: string[] = [];
    pendingList: string[] = [];

    show: boolean = false;

    constructor(private readonly websocketService: WebsocketService) {
        this.websocketService.on(FriendEvent.UPDATE_SINGLE_FRIEND, (update?: FriendUpdate) => {
            if (update) this.updateFriend(update.username, update.isAdded);
        });
        this.websocketService.on(FriendEvent.REQUEST_IGNORED, (username?: string) => {
            if (username) this.requestIgnored(username);
        });
        this.websocketService.on(NotificationEvent.ADD_NOTIFICATION, (notification?: Notification) => {
            if (notification) this.addPendingStatus(notification.senderUsername);
        });
        this.websocketService.on(FriendEvent.ADD_NOT_FRIEND, (username?: string) => {
            if (username) this.notFriendList.push(username);
        });
        this.getFriendsEvent();
    }

    ngOnDestroy() {
        this.websocketService.removeAllListeners(FriendEvent.UPDATE_SINGLE_FRIEND);
        this.websocketService.removeAllListeners(FriendEvent.REQUEST_IGNORED);
    }

    toggleShow(isOff: boolean = false) {
        if (isOff) this.show = false;
        else this.show = !this.show;
    }

    async getFriendsEvent() {
        const lists = await this.websocketService.sendWithAck<undefined, FriendLists>(FriendEvent.GET_FRIENDS);
        if (lists) {
            this.friendList = lists.friends;
            this.notFriendList = lists.notFriends;
            this.pendingList = lists.pending;
        }
    }

    removeFriendEvent(username: string) {
        this.websocketService.send<string>(FriendEvent.REMOVE_FRIEND, username);
    }

    respondToFriendRequest(username: string, isAccepted: boolean, notificationId: string) {
        this.websocketService.send<FriendRequestResponse>(FriendEvent.FRIEND_REQUEST_RESPONSE, { senderName: username, isAccepted, notificationId });
        const index = this.pendingList.indexOf(username);
        if (index !== NOT_FOUND_INDEX) this.pendingList.splice(index, 1);
        if (!isAccepted) this.notFriendList.push(username);
    }

    addPendingStatus(username: string) {
        this.pendingList.push(username);
        const index = this.notFriendList.indexOf(username);
        if (index !== NOT_FOUND_INDEX) this.notFriendList.splice(index, 1);
    }

    private updateFriend(username: string, isAdded: boolean) {
        if (isAdded) {
            this.friendList.push(username);
            const index1 = this.pendingList.indexOf(username);
            if (index1 !== NOT_FOUND_INDEX) this.pendingList.splice(index1, 1);
        } else {
            this.notFriendList.push(username);
            const index2 = this.friendList.indexOf(username);
            if (index2 !== NOT_FOUND_INDEX) this.friendList.splice(index2, 1);
        }
    }

    private requestIgnored(username: string) {
        this.notFriendList.push(username);
        const index = this.pendingList.indexOf(username);
        if (index !== NOT_FOUND_INDEX) this.pendingList.splice(index, 1);
    }
}
